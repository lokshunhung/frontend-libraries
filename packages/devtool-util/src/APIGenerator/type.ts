export interface APIDefinition {
    services: ServiceDefinition[];
    types: TypeDefinition[];
}

export interface ServiceDefinition {
    name: string;
    operations: Operation[];
}

export interface Operation {
    name: string;
    method: "DELETE" | "GET" | "POST" | "PUT";
    path: string;
    pathParams: {name: string; type: string}[];
    responseType: string;
    requestType: null | string;
}

export interface TypeDefinitionFieldConstraintSize {
    min: number;
    max: number;
}

export interface TypeDefinitionFieldConstraints {
    notNull: boolean | null;
    notBlank: boolean | null;
    min: number | null;
    max: number | null;
    size: TypeDefinitionFieldConstraintSize | null;
    pattern: string | null;
}

export type TypeParams = string[] | null;

export interface TypeDefinitionField {
    name: string;
    type: string;
    typeParams: TypeParams;
    constraints: TypeDefinitionFieldConstraints;
}

export interface TypeEnumConstant {
    name: string;
    value: string;
}

export enum JavaType {
    String = "String",
    Boolean = "Boolean",
    Integer = "Integer",
    Long = "Long",
    Double = "Double",
    BigDecimal = "BigDecimal",
    ZonedDateTime = "ZonedDateTime",
    LocalDate = "LocalDate",
    LocalDateTime = "LocalDateTime",
    LocalTime = "LocalTime",
    List = "List",
    Map = "Map",
}

export enum TypeDefinitionType {
    Enum = "enum",
    Bean = "bean",
}

export interface TypeDefinition {
    name: string;
    type: TypeDefinitionType;
    fields: TypeDefinitionField[] | null;
    enumConstants: TypeEnumConstant[] | null;
}

export interface PlatformConfig {
    ajaxFunction: string;
    ajaxFunctionImportStatement: string;
    typeFileImportPath: string;
}

export interface APIGeneratorOptions {
    metadataEndpointURL: string;
    typeFilePath: string;
    serviceFolderPath: string;
    platformConfig: PlatformConfig;
}

export type PlatformSpecificAPIGeneratorOptions = Omit<APIGeneratorOptions, "platformConfig">;
