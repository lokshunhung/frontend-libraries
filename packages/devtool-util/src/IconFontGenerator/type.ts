export interface WebIconFontGeneratorOptions {
    iconComponentDirectory: string;
    staticDirectory: string;
    // TODO/dion: do not remove this now, i will check again
    fontFamily?: string;
}

export interface AppIconFontGeneratorOptions {
    iconComponentFile: string;
    androidFontPath: string;
    iosFontPath: string;
}
