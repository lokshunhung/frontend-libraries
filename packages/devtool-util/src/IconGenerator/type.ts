export interface IconGeneratorOptions {
    iconComponentDirectory: string;
    svgDirectory: string;
    fontFamily?: string;
}
