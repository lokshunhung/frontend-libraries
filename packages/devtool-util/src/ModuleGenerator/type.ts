export interface ModuleGeneratorOptions {
    srcDirectory: string;
    templateDirectory: string;
    generateImportStatementForNewModuleState: (_: {moduleStateName: string; partialModulePath: string}) => string;
}

export type PlatformSpecificModuleGeneratorOptions = Omit<ModuleGeneratorOptions, "templateDirectory" | "generateImportStatementForNewModuleState">;
