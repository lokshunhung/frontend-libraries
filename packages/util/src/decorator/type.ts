export type AnyFunctionDecorator = (target: object, propertyKey: string, descriptor: TypedPropertyDescriptor<(...args: any[]) => any>) => TypedPropertyDescriptor<(...args: any[]) => any>;

export type VoidFunctionDecorator = (target: object, propertyKey: string, descriptor: TypedPropertyDescriptor<(...args: any[]) => void>) => TypedPropertyDescriptor<(...args: any[]) => void>;
