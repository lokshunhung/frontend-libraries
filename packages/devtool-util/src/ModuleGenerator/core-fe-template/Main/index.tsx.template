import {ReactUtil} from "@pinnacle0/util";
import React from "react";
import "./index.less";

export const Main = ReactUtil.memo("Main", () => {
    return <div />;
});
