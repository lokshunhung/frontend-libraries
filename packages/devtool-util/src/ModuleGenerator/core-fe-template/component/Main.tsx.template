import {ReactUtil} from "@pinnacle0/util";
import React from "react";

export const Main = ReactUtil.memo("Main", () => {
    return <div />;
});
