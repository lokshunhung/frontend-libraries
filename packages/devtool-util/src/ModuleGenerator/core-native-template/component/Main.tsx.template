import {ReactUtil} from "@pinnacle0/util";
import React from "react";
import {StyleSheet, View} from "react-native";

export const Main = ReactUtil.memo("Main", () => {
    return (
        <View style={styles.container}>
            <View />
        </View>
    );
});

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: "center",
        alignItems: "center",
    },
});
