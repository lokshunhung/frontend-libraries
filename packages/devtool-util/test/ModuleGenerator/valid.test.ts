import * as fs from "fs";
import * as path from "path";

const tmpDirectory = path.join(__dirname, "./__tmp__/valid");
const templatePath = path.join(__dirname, "./__tmp__/valid/module-template");
const moduleBasePath = path.join(__dirname, "./__tmp__/valid/modules");
const reduxStateTypePath = path.join(__dirname, "./__tmp__/valid/state.ts");

describe("ModuleGenerator class", () => {
    const fixtures = getFixtures();

    beforeAll(() => {
        fs.rmSync(tmpDirectory, {recursive: true, force: true});
        fixtures.forEach(_ => {
            const fixtureDirectory = path.dirname(_.path);
            if (!fs.existsSync(fixtureDirectory)) fs.mkdirSync(fixtureDirectory, {recursive: true});
            fs.writeFileSync(_.path, _.data, {encoding: "utf8"});
        });
    });

    afterAll(() => {
        // Comment the following line to see the temp files
        fs.rmSync(tmpDirectory, {recursive: true, force: true});
    });

    test("generate files", async () => {
        fixtures.forEach(_ => expect(fs.existsSync(_.path)).toBe(true));

        jest.doMock("yargs", () => ({
            argv: {_: {0: "common/new-feature"}},
        }));
        jest.doMock("../../src/PrettierUtil", () => ({
            PrettierUtil: {
                format: () => {},
            },
        }));

        const {ModuleGenerator} = jest.requireActual<typeof import("../../src/ModuleGenerator")>("../../src/ModuleGenerator");
        await new ModuleGenerator({
            moduleBasePath,
            reduxStateTypePath,
            templatePath,
        }).run();

        const expectedNewFiles = [
            path.join(moduleBasePath, "common/new-feature/component/Main.tsx"),
            path.join(moduleBasePath, "common/new-feature/hooks.ts"),
            path.join(moduleBasePath, "common/new-feature/index.ts"),
            path.join(moduleBasePath, "common/new-feature/type.ts"),
        ];
        // Check if each expected new file exists
        expectedNewFiles.forEach(path => {
            const exists = fs.existsSync(path) ? "exists" : "not-exists";
            expect(`${exists}: ${path}`).toBe(`exists: ${path}`);
        });
        // Create a snapshot for each new file
        expectedNewFiles.forEach(path => {
            expect(fs.readFileSync(path, {encoding: "utf8"})).toMatchSnapshot();
        });
    });
});

function getFixtures(): {path: string; data: string}[] {
    return [
        /* Template fixtures ------------------------------------------------ */
        {
            path: path.join(templatePath, "component/Main.tsx"),
            data: `// This file is generated from a template, but can be modified
import React from "react";

export const Main = () => {
    return <div />;
};
`,
        },
        {
            path: path.join(templatePath, "hooks.ts"),
            data: `// This file is generated from a template, but can be modified
import {RootState} from "app/type/state";
import {useSelector} from "react-redux";

export function use{1}<T>(fn: (state: RootState["app"]["{2}"]) => T): T {
    return useSelector((state: RootState) => fn(state.app.{2}));
}
`,
        },
        {
            path: path.join(templatePath, "index.ts"),
            data: `// This file is generated from a template, but can be modified
import {RootState} from "app/type/state";
import {Lifecycle, Module, register, SagaIterator} from "core-native";
import {Main} from "./component/Main";
import {State} from "./type";

class {1} extends Module<RootState, "{2}"> {}

const module = register(new {1}("{2}"), initialState);
export const actions = module.getActions();
export const MainComponent = module.attachLifecycle(Main);
`,
        },
        {
            path: path.join(templatePath, "type.ts"),
            data: `// This file is generated from a template, but can be modified
export interface State {}
`,
        },

        /* Source fixtures -------------------------------------------------- */
        {
            path: reduxStateTypePath,
            data: `//
import {State as OldFeatureState} from "./modules/common/old-feature";
import {State} from "core-fe";

export interface RootState extends State {
    app: {
        oldFeature: OldFeatureState;
    };
}
`,
        },
        {
            path: path.join(moduleBasePath, "common/old-feature/types.ts"),
            data: `//
export interface State {}
`,
        },
    ];
}
