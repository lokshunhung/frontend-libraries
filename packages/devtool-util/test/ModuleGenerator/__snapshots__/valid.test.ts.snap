// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleGenerator class generate files 1`] = `
"// This file is generated from a template, but can be modified
import React from \\"react\\";

export const Main = () => {
    return <div />;
};
"
`;

exports[`ModuleGenerator class generate files 2`] = `
"// This file is generated from a template, but can be modified
import {RootState} from \\"app/type/state\\";
import {useSelector} from \\"react-redux\\";

export function useNewFeatureState<T>(fn: (state: RootState[\\"app\\"][\\"newFeature\\"]) => T): T {
    return useSelector((state: RootState) => fn(state.app.newFeature));
}
"
`;

exports[`ModuleGenerator class generate files 3`] = `
"// This file is generated from a template, but can be modified
import {RootState} from \\"app/type/state\\";
import {Lifecycle, Module, register, SagaIterator} from \\"core-native\\";
import {Main} from \\"./component/Main\\";
import {State} from \\"./type\\";

class NewFeatureModule extends Module<RootState, \\"newFeature\\"> {}

const module = register(new NewFeatureModule(\\"newFeature\\"), initialState);
export const actions = module.getActions();
export const MainComponent = module.attachLifecycle(Main);
"
`;

exports[`ModuleGenerator class generate files 4`] = `
"// This file is generated from a template, but can be modified
export interface State {}
"
`;
